Recycle bin module

ledrypotato opened the pull request:
## Description

This PR adds the Recycle Bin module that will list files in the Recycle Bin. It will parse the associated metadata files in the Recycle Bin to display the "Original Location" of the deleted file, this can give a good indication whether or not the file is interesting or not. You can also download files using the module options.

I have a few things on my TODO list that I will implement as soon as I can as well as update the NetExec documentation.

Feel free to make any comments on the implementation/development as it's not my speciality 😆!

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

## How Has This Been Tested?

I have tested this module against my local Windows 11 VM (Microsoft Windows 11 Enterprise Evaluation) from my Kali VM running Python 3.12.8.

## Screenshots:

Here is an example of deleted files in the Recycle Bin:

![Pasted image 20250321181836](https://github.com/user-attachments/assets/507fd1f5-00ac-4204-87f4-41499a23b277)

We can list files in the Recycle Bin with this command:

```bash
poetry run nxc smb 192.168.1.42 -u user -p potato -M recycle_bin
```

![Pasted image 20250321180724](https://github.com/user-attachments/assets/c2d02f30-ed7a-437f-bce0-a1c4d3fd76d4)

or download them with this command:

```bash
poetry run nxc smb 192.168.1.42 -u user -p potato -M recycle_bin -o DOWNLOAD=true
```
![Pasted image 20250321183124](https://github.com/user-attachments/assets/e6d79d3c-33ad-49de-85d0-28809c9e664c)

## Checklist:

- [x] I have ran Ruff against my changes (via poetry: `poetry run python -m ruff check . --preview`, use `--fix` to automatically fix what it can)
- [x] I have added or updated the tests/e2e_commands.txt file if necessary
- [ ] New and existing e2e tests pass locally with my changes
- [x] My code follows the style guidelines of this project (should be covered by Ruff above)
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (PR here: https://github.com/Pennyw0rth/NetExec-Wiki/pull/39)

## TODO:

- [x] Implement the display of file deletion time to know when the file was deleted (this information should be in the metadata file but I couldn't parse it correctly)
- [x] Handle directories in the Recycle Bin as well as single files
- [x] Specify what files you want to download as a module option filter

ledrypotato committed:
initial Recycle Bin module

ledrypotato committed:
add e2e_commands

ledrypotato committed:
cleanup with Ruff

NeffIsBack commented:
Thanks for the PR!

If you still want to work on that module it's best to turn the PR into a "DRAFT", so that we know it is still worked on :)

ledrypotato committed:
handled retrieving deletion time + filter for file download

ledrypotato committed:
handle recursive folders as best as i can

ledrypotato committed:
cleanup with Ruff

ledrypotato committed:
add TODOs

ledrypotato committed:
add e2e_commands

ledrypotato commented:
I have made all the changes that I had planned and also created a PR for the associated documentation for the NetExec wiki.

ledrypotato commented:
Any updates?

NeffIsBack commented:
Hi,
to be honest i lost a bit track of all currently open PRs. 
A few minutes ago i just merged #https://github.com/Pennyw0rth/NetExec/pull/463 which has the same purpose. However, your PR contains a few pretty cool features that the module from @Dfte does not. Could you integrate your Pull Request into the current recycle bin module?

ledrypotato commented:
Hey, oh I didn't even realise there was a previous module for the Recycle Bin by @Dfte 😆 !
I will have to take a look and see how easy it is to integrate into that.

ledrypotato commented:
> @ledrypotato have you had a chance to look at the other module and see if you can integrate anything?

Hey, not yet unfortunately. I should have some more time beginning of next month.

ledrypotato committed:
Merged/moved my RecyleBin module into Defte's existing file.
I removed the registry key logic to fetch the username since it's already given in the metadata file ($I).

ledrypotato commented:
Hi, I finally had some time to look at this merge.

To be honest I moved most of my code that I had into Dfte's module. I removed the registry key logic to fetch the username since we can get that information from the metadata file in the Recycle Bin (files that start with $I).

- [ x ] I updated the e2e_commands.txt file
- [ x ] I ran ruff against the code

Here is an updated screenshot of the output when specifying that you want to download files (`-o DOWNLOAD=true`) with a filter on the file name (`FILTER=password`).

`poetry run nxc smb 192.168.1.42 -u user -p potato -M recyclebin -o DOWNLOAD=true FILTER=password`

<img width="1808" height="302" alt="image" src="https://github.com/user-attachments/assets/2ec8ebcf-95db-4f43-9494-fed8816028c3" />

Most of the code I originally posted remains the same. I felt it was easier to keep what I already had than try and merge everything. That said I did review the original code to check if things were better in it. Perhaps the `connection.spider` is more efficient, I will let you decide as you are more familiar.

Feel free to give any feedback on what should be changed/optimized.

ledrypotato commented:
Yeah the logic remains the same.
As a side note I personally prefer the naming convention with the underscore (recycle_bin) and it aligns with most other module names which are multiple words. I'll let you decide what suits best.

ledrypotato committed:
Merge branch 'main' into RecycleBin-module

ledrypotato commented:
Hey I assume this thread got lost will all the other open PRs. Let me know if anything needs changed before it can be merged.

NeffIsBack commented:
> Hey I assume this thread got lost will all the other open PRs. Let me know if anything needs changed before it can be merged.

More a struggle of good prioritization. But yeah, this PR has waited a long time now, i'll put it up the chain. Thanks for reminding.

Marshall-Hallenbeck commented on nxc/modules/recyclebin.py:
```diff
@@ -2,100 +2,135 @@
+        try:
+            connection.conn.getFile("C$", file_path, buf.write)
+        except Exception as e:
+            if "STATUS_FILE_IS_A_DIRECTORY" in str(e):
```
this if else makes no sense, it logs the same thing for both paths

Marshall-Hallenbeck reviewed (changes requested):
The download functionality reads the raw bytes and writes them to a file, but there's already functions to download for SMB instead of re-inventing the wheel. Why did you go with this method instead of using an existing one?

NeffIsBack committed:
Formatting

NeffIsBack committed:
Formatting

ledrypotato committed:
Merge branch 'Pennyw0rth:main' into RecycleBin-module

ledrypotato committed:
add helper function

ledrypotato committed:
replace recursion with spider method and handle nested folders better

ledrypotato committed:
Merge branch 'Pennyw0rth:main' into RecycleBin-module

ledrypotato commented:
For reference here is the outdated output:

Listing the files in each user's Recycle Bin:

<img width="2319" height="466" alt="image" src="https://github.com/user-attachments/assets/f8be8ca8-6b9c-4743-93c4-399368ec5d0f" />

Downloading a file with a specific name:

<img width="2531" height="612" alt="image" src="https://github.com/user-attachments/assets/7ee290dc-5cb2-4b6f-a253-def485789185" />

Also as a side note:

- [x] I have ran `ruff` against the code and applied the fixes
- [x] The e2e commands are correct

ledrypotato commented:
> If i don't review it in the coming 2-3 weeks please ping me :D

Whenever you get the chance 😄

NeffIsBack commented:
> > If i don't review it in the coming 2-3 weeks please ping me :D
> 
> 
> 
> Whenever you get the chance 😄 

Of course I forgot😅 Thanks for the ping, I will look at it the next days (hopefully tomorrow).

NeffIsBack commented on nxc/modules/recyclebin.py:
```diff
@@ -1,101 +1,181 @@
+        paths = connection.spider("C$", folder="$Recycle.Bin", regex=[r"(.*)"], silent=True)
+        filtered_paths = [path for path in paths if not path.endswith(self.false_positive)]
+        if not filtered_paths:
+            context.log.display("No files found in the Recycle Bin.")
```
I think that should be `.info()` so it doesn't spam too much when we scan large ranges

NeffIsBack commented on nxc/modules/recyclebin.py:
```diff
@@ -1,101 +1,181 @@
+    description = "Lists (and downloads) files in the Recycle Bin."
     supported_protocols = ["smb"]
     category = CATEGORY.CREDENTIAL_DUMPING
+    false_positive = (".", "..", "desktop.ini", "S-1-5-18")
```
Any reason why Default/Public etc are not included in the list? Do they not have a folder there?

NeffIsBack commented on nxc/modules/recyclebin.py:
```diff
@@ -1,101 +1,181 @@
 
-            for sid_directory in connection.conn.listPath("C$", "$Recycle.Bin\\*"):
+        paths = connection.spider("C$", folder="$Recycle.Bin", regex=[r"(.*)"], silent=True)
+        filtered_paths = [path for path in paths if not path.endswith(self.false_positive)]
```
Maybe we should do the filtering before the spidering, otherwise we waste unnecessary runtime (potentially a lot if these have large dirs)

NeffIsBack commented on nxc/modules/recyclebin.py:
```diff
@@ -1,101 +1,181 @@
+
+        remote_full_paths = [f"C$/{path}" for path in filtered_paths]
+
+        # extract the SID as the third component of the path (e.g. C$/$Recycle.Bin/SID/filename)
+        sid_dirs = []
+        for path in remote_full_paths:
+            sid_dir = path.split("/")[2]
+            if sid_dir not in sid_dirs:
+                sid_dirs.append(sid_dir)
+
+        # extract the full file path after the SID (e.g. filename or $Rfolder/filename) and remove empty values
+        remote_file_paths = ["/".join(path.split("/")[3:]) for path in remote_full_paths if "/".join(path.split("/")[3:])]
+
+        # parse the remote_files_paths and create a mapping of SID to the files that belong to that SID directory
+        sid_to_files = {}
+        for path in remote_full_paths:
+            sid_dir = path.split("/")[2]
+            if sid_dir not in sid_to_files:
+                sid_to_files[sid_dir] = []
+            sid_to_files[sid_dir].append(path)
+
+        context.log.display(f"Found {len(remote_file_paths)} files in the Recycle Bin across {len(sid_dirs)} user directories. Processing files...")
+
+        # check for each key in sid_to_files, if the value is a list with only 1 element and if that element is the same as the SID directory then we can skip processing that SID directory because it means that there are no files in that SID directory
+        for sid_dir, files in sid_to_files.items():
+            if len(files) == 1 and files[0].split("/")[2] == sid_dir:
+                context.log.display(f"No files found in C$\\$Recycle.Bin\\{sid_dir}. Skipping...")
+            else:
+                context.log.highlight(f"Processing directory: C$\\$Recycle.Bin\\{sid_dir} with {len(files) - 1} file(s)...")
+                for file in files:
+                    context.log.debug(f"Found file: {file}")
```
This looks overly complicated, what is that trying to achieve? Can't we just do a log statement in the primary looping logic down below?

Also, it looks to me like currently we are only looping over the last iterated `sid_dir` as this is set in the loop above and appended to the path in the main for loop down below. I haven't actually tested it yet, but does this indeed iterate over all sids? Maybe I am missing something here tho.

NeffIsBack commented on nxc/modules/recyclebin.py:
```diff
@@ -1,101 +1,181 @@
+
+            # Process Metadata files ($I) to extract original path, deletion time and size
+            if remote_file_path.startswith("$I"):
+                export_path = join(NXC_PATH, "modules", "recyclebin")
+                makedirs(export_path, exist_ok=True)
```
Just do that once at the start of the module execution, otherwise this will be checked in every loop which is unnecessary and in addition is duplicate to the logic in L167-L169

NeffIsBack commented on nxc/modules/recyclebin.py:
```diff
@@ -1,101 +1,181 @@
+
+                with open(dest_path, "wb+") as f:
+                    connection.conn.getFile("C$", remote_full_path, f.write)
+                f.close()
```
You don't need to close the file, this is already done by the context manager (`with open...`)

NeffIsBack commented on nxc/modules/recyclebin.py:
```diff
@@ -1,101 +1,181 @@
+                            original_path = match.group(1)
+                            metadata_map[remote_file_path.replace("$I", "")] = original_path
+                            size_map[remote_file_path.replace("$I", "")] = file_size_raw
+                            size_display = f"{file_size_raw // 1024}KB" if file_size_raw >= 1024 else f"{file_size_raw}B"
```
We should probably move the `convert_size` function from nfs.py to nxc/helpers/misc, add a doc string and use it in nfs and here. See: https://github.com/Pennyw0rth/NetExec/blob/1884cd30a92624bd33bf8d50c14297afa4dc6ad0/nxc/protocols/nfs.py#L754-L761

NeffIsBack commented on nxc/modules/recyclebin.py:
```diff
@@ -1,101 +1,181 @@
-                    if "ERROR_FILE_NOT_FOUND" in str(e):
-                        continue
-                    else:
-                        context.log.fail(f"Error opening {sid_directory.get_longname()} on host {connection.host} because of {e}")
+                    remove(dest_path)
+                    context.log.debug(f"Deleted metadata file: {dest_path}")
+                except Exception:
+                    context.log.debug(f"Could not delete metadata file: {dest_path}")
```
If we remove the file anyway we should probably use BytesIO, so we can skip all that creating dir, writing to disk etc. logic. See: https://github.com/Pennyw0rth/NetExec/blob/1884cd30a92624bd33bf8d50c14297afa4dc6ad0/nxc/modules/gpp_password.py#L46-L47

NeffIsBack replied:
God, that diff looks crazy due to the "removal" detected by git. This belongs only to the 4 addition lines

NeffIsBack commented on nxc/modules/recyclebin.py:
```diff
@@ -1,101 +1,181 @@
+                    except Exception as e:
+                        context.log.debug(f"Could not get file size for {remote_file_path} via listPath: {e}")
+
+                size_display = f"{file_size_raw // 1024}KB" if file_size_raw and file_size_raw >= 1024 else f"{file_size_raw}B" if file_size_raw else "unknown"
```
See above

NeffIsBack commented on nxc/modules/recyclebin.py:
```diff
@@ -1,101 +1,181 @@
+                        with open(path, "wb") as f:
+                            connection.conn.getFile("C$", remote_full_path, f.write)
+                        context.log.success(f"Recycle Bin file {remote_file_path} written to: {path}")
+                        f.close()
```
See above

NeffIsBack commented on nxc/modules/recyclebin.py:
```diff
@@ -1,101 +1,181 @@
+                    path = abspath(join(export_path, local_filename))
+                    makedirs(export_path, exist_ok=True)
+
+                    if (len(remote_file_path) <= 8):  # if it's a folder we can't download it and we should skip it
+                        context.log.debug(f"Skipping download of {remote_file_path} because it appears to be an empty directory.")
```
Why is it a folder when the len is <= 8? Isn't that a check in the name of the dir/file?

NeffIsBack reviewed (changes requested):
Hey, sorry for the delay, tight week. Here is the review tho.

Looking at the structure, we are downloading $I and $R files, mapping metadata information from $I to $R. However, with the current structure this would mean if we retrieve the $R file before we retrieved $I we would miss out on the metadata information. 

What about we are just looking for $R and then try to download the respective metadata file? Then we wouldn't have to rely on some internal metadata mapping where everything could exist or not. Thoughts?

ledrypotato commented:
Thanks for the in-depth review. I won't have time to check the review this week and I won't be here for the next few weeks so it will have to wait a bit 😄 .

NeffIsBack commented:
> Thanks for the in-depth review. I won't have time to check the review this week and I won't be here for the next few weeks so it will have to wait a bit 😄 .

No worries, we are not in a rush :D
